src/Symfony/Component/Messenger/Worker.php

```diff
@@ -60,21 +61,33 @@ public function __construct(ReceiverInterface $receiver, MessageBusInterface $bu
     /**
      * Receive the messages and dispatch them to the bus.
      */
-    public function run()
+    public function run(callable $onHandledCallback = null): void
     {
         if (\function_exists('pcntl_signal')) {
             pcntl_signal(SIGTERM, function () {
-                $this->receiver->stop();
+                $this->stop();
             });
         }
 
-        $this->receiver->receive(function (?Envelope $envelope) {
+        $handled = function (?Envelope $envelope) use ($onHandledCallback) {
+            if (\function_exists('pcntl_signal_dispatch')) {
+                pcntl_signal_dispatch();
+            }
+
+            if (null !== $onHandledCallback) {
+                $onHandledCallback($envelope);
+            }
+        };
+
+        while (false === $this->shouldStop) {
+            $envelope = $this->receiver->get();
+
             if (null === $envelope) {
-                if (\function_exists('pcntl_signal_dispatch')) {
-                    pcntl_signal_dispatch();
-                }
+                $handled(null);
+
+                usleep(1000000);
```

Comment:
Two changes here:

1) I changed from .2 seconds to 1 second. Waiting only 200ms between when there is not a message is quite rapid. For reference, in Laravel, this defaults to 3 and is configurable on the command/worker.